source/common/network/utility.cc

```diff
@@ -66,13 +66,37 @@ std::string hostFromUrl(const std::string& url, absl::string_view scheme,
     throw EnvoyException(fmt::format("expected {} scheme, got: {}", scheme_name, url));
   }
 
-  const size_t colon_index = url.find(':', scheme.size());
+  const size_t bracket_begin_index = url.find('[', scheme.size());
 
-  if (colon_index == std::string::npos) {
-    throw EnvoyException(absl::StrCat("malformed url: ", url));
-  }
+  // If there is no '[', it means it is ipv4 address, otherwise it is ipv6 address.
+  if (bracket_begin_index == std::string::npos) {
```

Comment:
Hm. I'm slightly apprehensive about doing all this manual URL parsing since, as this PR points out, it's easy to get wrong. Would it be possible to instead use the google URL library which is already an envoy dependency?

Using this, I think it should be roughly as simple as:
```
return GURL(url).host();
```
(Modulo error checking, etc). For example, here's the GURL code which handles IP address canonicalization:

https://quiche.googlesource.com/googleurl/+/refs/heads/master/url/url_canon_ip.cc

Reply:
Yea, let me take a look at. Thanks for the review

Reply:
@RyanTheOptimist I did a test, seems `GURL` doesn't support the scheme of `tcp` and `udp` .

https://quiche.googlesource.com/googleurl/+/refs/heads/master/url/url_constants.cc

Reply:
Should we add those schemes to that library? cc. @htuch @yanavlasov

Reply:
Yes, that seems the cleanest thing, or at least make it easy to add additional schemes.

Reply:
Do you mean contribute that back to chromium commuinity?

Reply:
We unfortunately can not use GURL directly because of ICU dependency. But you should be able to use this class https://github.com/envoyproxy/envoy/blob/05bdb1d6850da7b00f1ee3b0f4d8bc2f3131f7c9/source/common/http/utility.h#L144 that should deal with the IPv6 hostnames correctly.

Reply:
Oh right, we have an independent Url class still, that's much better, we can hide the details later as needed if we switch to GURL there.

Reply:
thanks for the feedback, let me update